File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        boolean[][] vis = new boolean[m][n]; 
        boolean result = false;
        for (int i=0;i<m;i++) {
            for (int j=0;j<n;j++) {
                if(board[i][j] == word.charAt(0)) {
                    vis[i][j] = true;
                    result = dfs(board, vis, i, j, m, n, 1, word);
                    vis[i][j] = false;

                    if (result) {
                        return true;
                    }
                }
            }
        }

        return false;
    }

    private void clean(boolean[][] vis) {
        int m = vis.length;
        for (int i =0;i<m;i++) {
            Arrays.fill(vis[i], false);
        }
    }

    private boolean dfs(char[][] board, boolean[][] vis, int i, int j, int m, int n, int k, String word) {
        if (i >=m || i<0 || j>=n || j<0) {
            return false;
        }

        if (k >= word.length()) {
            return true;
        }

        boolean isPossible = false;

        if (i+1<m && !vis[i+1][j] && board[i+1][j] == word.charAt(k)) {
            vis[i+1][j] = true;
            isPossible |= dfs(board, vis, i + 1, j, m, n, k+1, word);
            vis[i+1][j] = false;
        }

        if (i-1>=0 && !vis[i-1][j] && board[i-1][j] == word.charAt(k)) {
            vis[i-1][j] = true;
            isPossible |= dfs(board, vis, i -1, j, m, n, k+1, word);
            vis[i-1][j] = false;
        }

        if (j+1<n && !vis[i][j+1] && board[i][j+1] == word.charAt(k)) {
            vis[i][j+1] = true;
            isPossible |= dfs(board, vis, i, j+1, m, n, k+1, word);
            vis[i][j+1] = false;
        }

        if (j-1>=0 && !vis[i][j-1] && board[i][j-1] == word.charAt(k)) {
            vis[i][j-1] = true;
            isPossible |= dfs(board, vis, i, j-1, m, n, k+1, word);
            vis[i][j-1] = false;
        }

        return isPossible;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        def dfs(x, y, m, n, board, idx, word) -> bool:
            if idx >= len(word):
                return True

            if x < 0 or x >= m or y < 0 or y >= n or board[x][y] != word[idx]:
                return False

            temp = board[x][y]
            board[x][y] = ''

            dr = [[0, 1], [0,-1], [1, 0], [-1, 0]]

            res = False

            for d in dr:
                new_x = x + d[0]
                new_y = y + d[1]

                res = dfs(new_x, new_y, m, n, board, idx + 1, word)
                if res:
                    return True
                    
            
            board[x][y] = temp
            return False
        
        m = len(board)
        n = len(board[0])

        for idx in range(m):
            for jIdx in range(n):
                res = dfs(idx, jIdx, m, n, board, 0, word)
                if res:
                    return True
        return False



        